File: src/test/java/com/raychenon/leetcode/LongestCommonSubsequenceTest.java
package com.raychenon.leetcode;

import org.junit.BeforeClass;
import org.junit.Test;

import static junit.framework.TestCase.assertEquals;

/**
 * User: raychenon
 * Date: 9/5/2020
 * https://leetcode.com/problems/longest-common-subsequence/
 */
public class LongestCommonSubsequenceTest {

    private static LongestCommonSubsequence instance;

    @BeforeClass
    public static void setUp() {
        instance = new LongestCommonSubsequence();
    }

    @Test
    public void example1Test() {
        //  The longest common subsequence is "ace"
        assertLongestCommonSubsequenceMethods(3, "abcde", "ace");
    }

    @Test
    public void example2Test() {
        // The longest common subsequence is "abc"
        assertLongestCommonSubsequenceMethods(3, "abc", "abc");
    }

    @Test
    public void example3Test() {
        // There is no such common subsequence,
        assertLongestCommonSubsequenceMethods(0, "abc", "def");
    }

    private void assertLongestCommonSubsequenceMethods(int expected, String text1, String text2) {
        assertEquals(expected, instance.longestCommonSubsequenceDP(text1, text2));
        assertEquals(expected, instance.longestCommonSubsequenceDPreverse(text1, text2));
        assertEquals(expected, instance.longestCommonSubsequenceDPSpaceOptimized(text1, text2));
        assertEquals(expected, instance.longestCommonSubsequenceRecursive(text1, text2));
    }
}

File: src/test/kotlin/com/raychenon/kotlin/leetcode/LongestCommonSubsequenceTest.kt
package com.raychenon.kotlin.leetcode

import junit.framework.TestCase
import org.junit.Test

/**
 * User: raychenon
 * Date: 11/5/2020
 * https://leetcode.com/problems/longest-common-subsequence/
 */
class LongestCommonSubsequenceTest {


    @Test
    fun example1Test() {
        //  The longest common subsequence is "ace"
        assertLongestCommonSubsequenceMethods(3, "abcde", "ace")
    }

    @Test
    fun example2Test() {
        // The longest common subsequence is "abc"
        assertLongestCommonSubsequenceMethods(3, "abc", "abc")
    }

    @Test
    fun example3Test() {
        // There is no such common subsequence,
        assertLongestCommonSubsequenceMethods(0, "abc", "def")
    }

    @Test
    fun exampleJAVAIDTest() {
        // The longest common subsequence is "javaid"
        assertLongestCommonSubsequenceMethods(6, "javaaid", "javaid", true)
    }

    @Test
    fun exampleDNA2Test() {
        assertLongestCommonSubsequenceMethods(5, "actgattag", "gtgtgatcg", true)
    }

    private fun assertLongestCommonSubsequenceMethods(
        expected: Int,
        text1: String,
        text2: String,
        debug: Boolean = false
    ) {
        TestCase.assertEquals(expected, LongestCommonSubsequence.longestCommonSubsequenceDP(text1, text2, debug))
        TestCase.assertEquals(expected, LongestCommonSubsequence.longestCommonSubsequenceDP1Darray(text1, text2, debug))
        TestCase.assertEquals(expected, LongestCommonSubsequence.longestCommonSubsequenceDPreverse(text1, text2))
    }
}